docs/adr/0006-json-schemas-for-messaging.md

```diff
@@ -0,0 +1,92 @@
+use-case only and does not apply to JSON used in RESTful APIs.
+
+Tracking and minimizing drift between the accuracy of message format documentation and
+the underlying implementation is an error-prone process. In addition, JSON messages are
+frequently produced and consumed from multiple systems, and in multiple langauges. Ensuring
+validity of these messages frequently requires implementing validation logic in each
+language that consumes these JSON messages to ensure correctness for message handling.
+
+Adding a field to a message format is non-trivial, as every consumer must also update
+its own validation logic. Currently there are a variety of places where messages are consumed
+and produced, most producers do not validate the message they send as the consumer code is not
+always in the same service where the validation code typically resides.
+
+There is currently no versioning for our message formats, which also creates difficulties in
+knowing when validation logic must change and when existing flexibility is sufficient.
+
+## Decision Drivers
+
+- Accurate documentation via JSON-Schema doc generation
+- Validation logic generated from JSON-Schema to reduce manually written/maintained code
+- [Semantic Versioning][schemaver] to effectively communicate breaking vs. non-breaking message format changes
+
+## Considered Options
+
+1. Keep existing approach
+2. Use protobuf to format/validate/document message formats
+3. Use JSON-Schemas with SemVer for doc/validation of existing JSON message formats
+
+## Decision Outcome
+
+PLACEHOLDER
+
+### [option 1] Keep existing approach
+
+Currently our documentation drifts if a message format is tweaked and the docs are not updated. Ideally
+our code review process would catch this but sometimes code subleties are not caught that alter the
+message format. Validation changes can also fail to appear in the documentation, even though what is
+considered a valid message key/value should be documented.
+
+We have to occasionally duplicate `joi` schemas between repos, which should likely be moved to `fxa-shared`
+to benefit re-use, but this does not help non-Node projects consume our messages.
+
+- Pros
+  - No additional work is needed that we don't already do.
+- Cons
+  - Our documentation occasionally drifts.
+  - It's additional work to maintain accurate `joi` schemas spread throughout the code-base _or_
+  - It's additional work to consolidate our `joi` schemas in `fxa-shared`.
+
+### [option 2] Use protobuf
+
+Using protobuf would allow us to validate and document valid message formats that would also consume
+less space and be more performant to serialize/deserialize. Validation is built-in as invalid messages
+will not deserialize, and documenation can be generated based on the protobuf spec for a message.
+
+Protobuf tooling is widely available for most langauges.
+
+- Pros
+  - High throughput messaging could benefit from higher efficiency of serializing/deserializing of messages
```

Comment:
For the event broker, its not that important, our message throughput isn't terribly high. We're maybe a few per second vs. the hundreds or thousands per second that would make this a bigger priority.

Reply:
There are messages *to* the event broker, and there are messages *from* the event broker. It might be a bit much to ask of service providers to set up protobuf to consume message coming *from* the event broker.

Reply:
Agreed, I definitely think JSON-Schema is the way to go.